uroute/url.py

```diff
@@ -67,31 +67,37 @@ def clean_url(self, url):
             if not re.match(provider['urlPattern'], url, re.IGNORECASE):
                 continue
 
+            # If any exceptions are matched, this provider is skipped
             if any(
                 re.match(exc, url, re.IGNORECASE)
-                for exc in provider['exceptions']
+                for exc in provider.get('exceptions', [])
             ):
                 continue
 
-            for redir in provider['redirections']:
+            # If redirect found, recurse on target (only once)
+            for redir in provider.get('redirections', []):
                 match = re.match(redir, url, re.IGNORECASE)
                 try:
                     if match and match.group(1):
-                        return unquote(match.group(1))
+                        if loop:
+                            return self.clean_url(unquote(match.group(1)),
```

Comment:
After processing a redirect URL, send it back through the cleaning process to have any unwanted elements removed from the resulting link. Set the loop flag to false to guard against the possibility of a maliciously crafted link causing infinite recursion.